### PR TITLE
[Naitei][Ecommerce]  [User]  transaction for payment and concurrency

### PR DESCRIPTION
**Pull request by SuKay712**
## Related Tickets
https://edu-redmine.sun-asterisk.vn/issues/79581

## WHAT (optional)
- Change number items `completed/total` in admin page.

## HOW
- I edit js file, inject not_vary_normal items in calculate function.

## WHY (optional)
- Because in previous version - number just depends on `normal` items. But in new version, we have `state` and `confirm_state` depends on both `normal` + `not_normal` items.

## Evidence (Screenshot or Video)
https://drive.google.com/file/d/1z4PPerrMhUgUqI06LI80d1gbQ6xPmxbS/view?usp=sharing

### CODE REVIEW

**Comment by SuKay712** [comment]
ready

File: app/controllers/checkout_controller.rb [hunk 1]
```diff
@@ -2,19 +2,19 @@ class CheckoutController < ApplicationController
   skip_before_action :verify_authenticity_token
   before_action :load_current_user_cart
   def create
-    @bill = Bill.new bill_params
+    @bill = current_user.bills.build bill_params
     if @bill.save
       create_stripe_session @cart_details
       transfer_data
-      current_user.send_bill_info @bill
+      # current_user.send_bill_info @bill
```
Comment (ghost) [inline app/controllers/checkout_controller.rb]: không dùng thì xóa đi em

File: app/controllers/webhooks_controller.rb [hunk 2]
```diff
@@ -37,18 +37,44 @@ def handle_event event
...
+      end
+    rescue ActiveRecord::RecordInvalid, ActiveRecord::Rollback => e
+      handle_transaction_failure e
+    end
+  end
+
+  def update_bill_and_cart
+    @bill.update!(status: :wait_for_prepare, expired_at: nil)
+    @cart.update!(total: Settings.digit_0)
+    @cart.cart_details.try(:destroy_all)
```
Comment (ghost) [inline app/controllers/webhooks_controller.rb]: bỏ bớt () đi em

File: config/settings.yml [hunk 3]
```diff
@@ -39,3 +39,4 @@ digit_5 : 5
 init_value_add_cart: 1
 init_min_add_cart: 1
 dicount_percentage_convert: 100
+out_of_stocks: "Out of stocks"
```
Comment (ghost) [inline config/settings.yml]: Cái này đưa vào i18n chứ nhỉ

**Review by ghost — COMMENTED** [review]
cmt

File: config/settings.yml (hunk 3 above)
Comment (SuKay712) [reply config/settings.yml]: dạ sửa thành enum luôn rồi ạ

**Comment by SuKay712** [comment]
ready

**Review by ghost — COMMENTED** [review]
merged

**Pull request merged**